datafusion/functions/src/core/mod.rs

```diff
@@ -39,14 +42,68 @@ make_udf_function!(getfield::GetFieldFunc, GET_FIELD, get_field);
 make_udf_function!(coalesce::CoalesceFunc, COALESCE, coalesce);
 
 // Export the functions out of this package, both as expr_fn as well as a list of functions
-export_functions!(
-    (nullif, arg_1 arg_2, "returns NULL if value1 equals value2; otherwise it returns value1. This can be used to perform the inverse operation of the COALESCE expression."),
-    (arrow_cast, arg_1 arg_2, "returns arg_1 cast to the `arrow_type` given the second argument. This can be used to cast to a specific `arrow_type`."),
-    (nvl, arg_1 arg_2, "returns value2 if value1 is NULL; otherwise it returns value1"),
-    (nvl2, arg_1 arg_2 arg_3, "Returns value2 if value1 is not NULL; otherwise, it returns value3."),
-    (arrow_typeof, arg_1, "Returns the Arrow type of the input expression."),
-    (r#struct, args, "Returns a struct with the given arguments"),
-    (named_struct, args, "Returns a struct with the given names and arguments pairs"),
-    (get_field, arg_1 arg_2, "Returns the value of the field with the given name from the struct"),
-    (coalesce, args, "Returns `coalesce(args...)`, which evaluates to the value of the first expr which is not NULL")
-);
+pub mod expr_fn {
```

Comment:
It is not possible to create an expr_fn that takes a `Vec<Expr>` using the `export_functions` macro, so follow the model @Omega359  used in the math module https://github.com/apache/datafusion/blob/af3950698879a1e159fbe579bd433a0d5ce1c7cb/datafusion/functions/src/math/mod.rs#L78-L270

Reply:
I think it is possible to take `Vec` like what functions-array macro does

```rust
($UDF:ty, $EXPR_FN:ident, $DOC:expr , $SCALAR_UDF_FN:ident) => {
        paste::paste! {
            // "fluent expr_fn" style function
            #[doc = $DOC]
            pub fn $EXPR_FN(arg: Vec<Expr>) -> Expr {
                Expr::ScalarFunction(ScalarFunction::new_udf(
                    $SCALAR_UDF_FN(),
                    arg,
                ))
            }
```

Reply:
I spent a whole day trying to make it work as well and got lost in the depths of macro programming. I'm not sure it's worth the effort of trying to get it to work tbh - the alternative is not exactly hard work.

Reply:
https://github.com/apache/datafusion/blob/e3487eee1365a37b8651bf5a393ae8d2cc16e653/datafusion/functions-array/src/macros.rs#L47-L109

In array macro, we support two syntaxes, which is surprisingly easy

Reply:
The difference is that in this case there is a single macro call to `export_functions!` that has a single (variadic) argument

In the array cases there are multiple invocations of the macro, and the different invocation has different parameters

Reply:
I was thinking of https://github.com/alamb/datafusion/pull/19.

However, it does not reuse the macro `export_functions` that accepts multiple functions with one macro call.

`export_functions_single` accepts a single function with each macro call.

Since this requires introducing another macro, it is also not an ideal solution.

I'm fine to move on with the current PR.

Reply:
https://users.rust-lang.org/t/macro-repetition-with-multiple-rules/110816/2?u=jayzhan

I got a slight better solution from rust forum. To anyone that is interested in

Reply:
Thank you @jayzhan211  -- I will give it a try

Reply:
I couldn't make this work (though I didn't try very hard) so I filed https://github.com/apache/datafusion/issues/10397 as a follow on task for this

Reply:
I think we can merge this first and release 38.